src/Simulation/QCTraceSimulator/QubitAvailabilityTimeTracker.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.Quantum.Simulation.QCTraceSimulatorRuntime
+{
+    internal class QubitAvailabilityTimeTracker
+    {
+        private double DefaultAvailabilityTime = 0;
+        private List<double> QubitAvailableAt;
+        private long MaxQubitId = -1;
+
+        internal QubitAvailabilityTimeTracker(int initialCapacity, double defaultAvailabilityTime)
+        {
+            DefaultAvailabilityTime = defaultAvailabilityTime;
+            QubitAvailableAt = new List<double>(initialCapacity);
+        }
+
+         internal double this[long index]
+         {
+            get
+            {
+                if (index < QubitAvailableAt.Count)
+                {
+                    return QubitAvailableAt[(int)index];
+                }
+                return DefaultAvailabilityTime;
+            }
+            set
+            {
+                if (index == QubitAvailableAt.Count)
+                {
+                    QubitAvailableAt.Add(value);
+                    return;
+                }
+                else if (index >= int.MaxValue)
+                {
+                    throw new IndexOutOfRangeException("Too many qubits to track.");
+                }
+                else if (index > QubitAvailableAt.Count)
+                {
+                    QubitAvailableAt.AddRange(Enumerable.Repeat(DefaultAvailabilityTime, (int)index - QubitAvailableAt.Count + 1));
+                }
+                QubitAvailableAt[(int)index] = value;
+            }
+        }
+
+        internal long GetMaxQubitId()
+        {
+            return MaxQubitId;
+        }
+
+        internal void MarkQubitIdUsed(long qubitId)
```

Comment:
Is it possible that this relies on exactly knowledge about the inner workings of the qubit manager, i.e. it relies on how the qubit manager currently allocates qubits?

Reply:
It does not rely on qubit manager inner workings in general. It only relies on the fact that the same id will be assigned to the same underlying qubit that qubit manager manages. I.e. If a program does allocate-release-allocate-release, and qubit manager assigns the same underlying qubit for the second allocate-release cycle, it will have same id as in the first. This information (detecting reuse of qubits) is crucial for computing compatible depth and width. However, qubit _object_ generated for the second round may be different.

Reply:
My concern was that it might rely e.g. on consecutiveness of the numbering. But looking closer I believe this is not the case, correct?

Reply:
It does not rely on consecutiveness. But it's performance relies on IDs being a dense set.